internal/experiment/openvpn/targets.go

```diff
@@ -1,49 +1,170 @@
 package openvpn
 
 import (
+	"context"
 	"fmt"
 	"math/rand"
-	"net"
+	"slices"
+	"time"
+
+	"github.com/ooni/probe-cli/v3/internal/legacy/netx"
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
 )
 
-const defaultOpenVPNEndpoint = "openvpn-server1.ooni.io"
-
-// this is a safety toggle: it's on purpose that the experiment will receive no
-// input if the resolution fails. This also implies that we have no way of knowing if this
-// target has been blocked at the level of DNS.
-// TODO(ain,mehul): we might want to try resolving with other techniques (DoT etc),
-// and perhaps also transform DNS failure into a specific failure of the experiment, not
-// a skip.
-// TODO(ain): update the openvpn spec to reflect the CURRENT state of delivering the targets.
-func resolveTarget(domain string) (string, error) {
-	ips, err := net.LookupIP(domain)
-	if err != nil {
-		return "", err
+// defaultOONIEndpoints is the array of hostnames that will return valid
+// endpoints to be probed. Do note that this is a workaround for the lack
+// of a backend service; if you maintain this experiment in the future please
+// feel free to remove this workaround after the probe-services for distributing
+// endpoints has been deployed to production.
+var defaultOONIEndpoints = []string{
+	"a.composer-presenter.com",
+	"a.goodyear2dumpster.com",
+}
+
+// maxDefaultOONIAddresses is how many IPs to use from the
+// set of resolved IPs.
+var maxDefaultOONIAddresses = 3
+
+// sampleN takes max n elements sampled ramdonly from the array a.
+func sampleN(a []string, n int) []string {
```

Comment:
This function can probably be optimized by shuffling the list and return the first N elements from it instead of having to keep track of which indexes you picked as this might have a not so predictable runtime.

Reply:
fixed. 

```
goos: linux
goarch: amd64
pkg: github.com/ainghazal/test
cpu: AMD Ryzen 7 5800X 8-Core Processor
                     │ results.txt │
                     │   sec/op    │
SampleFunc1_Small-16   5.322µ ± 1% <- baseline
SampleFunc2_Small-16   5.286µ ± 1% <- new version
SampleFunc3_Small-16   5.482µ ± 1%
geomean                5.362µ
```